src/nccl_ofi_net.c

```diff
@@ -1414,40 +1416,73 @@ static ncclResult_t ofi_connect(int dev, void *handle, void **sendComm)
 	}
 
 	/* Send "connect" message to remote EP */
-	do {
-		/*
-		 * TODO: replace it with API of FI_INJECT type when most of
-		 * providers can support it, so that need for completion check
-		 * below can be lifted.
-		 */
-		rc = fi_tsend(sComm->local_ep, (void *)&local_ep_addr,
-			      MAX_EP_ADDR, NULL, sComm->remote_ep,
-			      sComm->tag | ~max_tag, &req->ctx);
-		if (rc == 0)
-			break;
+	/* do not wait for completion in case of sending to self */
+	if (memcmp(local_ep_addr, remote_ep_addr, namelen) == 0) {
+		static char buffer_to_self[MAX_EP_ADDR];
+		NCCL_OFI_TRACE(NCCL_NET, "Connect to self");
+		if (request_to_self != NULL) {
+			NCCL_OFI_WARN("Another connect to self request already in progress");
+			ret = ncclSystemError;
+			goto error;
+		}
+		memcpy(buffer_to_self, local_ep_addr, MAX_EP_ADDR);
+		rc = fi_tsend(sComm->local_ep, buffer_to_self,
+		              MAX_EP_ADDR, NULL, sComm->remote_ep,
+		              sComm->tag | ~max_tag, &req->ctx);
+		if (rc == 0) {
+			NCCL_OFI_TRACE(NCCL_NET, "Connect to self request completed");
+		}
 		else if (rc == -FI_EAGAIN) {
-			/*
-			 * Process completions so that you have enough
-			 * resources for sending connect message
-			 */
-			ret = nccl_ofi_progress(nccl_ofi_component[dev]);
-			if (OFI_UNLIKELY(ret != 0))
-				goto error;
+			NCCL_OFI_TRACE(NCCL_NET, "Connect to self request will be completed once receiver is ready");
```

Comment:
This basically means that the request to self wasn't delivered via `fi_tsend`. We simply created a request and saved it in the code for using at `accept` phase. If that's not a requirement then why do you have to call `fi_tsend` at all.

How do we ensure that we actually send the data using `fi_tsend`?

Reply:
> This basically means that the request to self wasn't delivered via `fi_tsend`. We simply created a request and saved it in the code for using at `accept` phase. If that's not a requirement then why do you have to call `fi_tsend` at all.

There is no such requirement, but plugin code transfers remote address via this fi_tsend, and it is the only way to differentiate "to self" accepts from "generic" ones. Please note that ofi_accept() does not use sendComm as a parameter, so there is no way to pass some boolean flag inside it.

> How do we ensure that we actually send the data using `fi_tsend`?

By successful fi_trecv() in ofi_accept(). For local communications it is overkill to check data transfer twice, so skip checking at sender side.

Reply:
How can we expect to receive a buffer from provider if the `tsend` ended up in `-EAGAIN`? From what I read from tsend man page (https://ofiwg.github.io/libfabric/v1.14.0/man/fi_tagged.3.html). `-EAGAIN` indicates to applications that they should re-try the send request and nothing has been sent actually.

Reply:
Reworked